### PR TITLE
[ENG-3865] Run Android E2E tests on EAS Build

### PR DESCRIPTION
**Pull request by dsokal**
# Why 

We've enabled nested virtualization for Android builds. Thanks to that, we can run Android emulator on EAS Build to run E2E tests.

# How

1. Cleaned up code from https://github.com/expo/eas-tests-example/pull/2 - the PR hasn't been reviewed:
  - Remove unused/leftover code.
  - Don't call `adapter.afterAll();` in `afterAll` hook in setup.js - https://github.com/expo/eas-tests-example/pull/3/files#diff-4d4a78dcedcf05fcf1221add78a46040125717a1b7aadf07f90669c026a8b487L12 I'm not sure why this was there. This doesn't seem to be necessary. Also, removing this line let me get rid of the `patch-package` module and the patch for `tempfile`. @douglowder 
  - Simplify/clean up code for opening app with development client - https://github.com/expo/eas-tests-example/pull/3/files#diff-77e4c4f47296231cbd04163d5739819172bee3aecf946ffb3b093959876375ee Instead of depending on fragile `const argparse = require('detox/src/utils/argparse');`, use the `DETOX_CONFIGURATION` env var to detect the platform and target. Introduce `EXPO_USE_UPDATES` for selecting bundle source (this is much cleaner).
  - Remove `e2e/testdeeplink.js` as I feel this doesn't belong to this repo.
2. Configure EAS Build to run E2E tests for Android:
  - Install `@config-plugins/detox` + run prebuild to generate `android/` (you can skip reviewing files in the directory as the files are auto-generated).
  - Add `android.debug` (I haven't been able to run tests with this config successfully; Detox is stuck at launching the app even though it is successfully launched in the emulator) and `android.release` configuration to `.detoxrc.json`. 
  - Install emulator system deps in `pre-install` hook + the pixel_4 emulator.
  - Run the emulator and then run Android tests in `on-success` hook. Running the emulator before running Detox seems to be fixing the problem where Detox is unable to launch the app for the first time 🤷 

# Test plan

`eas build -p all -e test` and see the builds finish successfully.

Android: https://expo.dev/accounts/dsokal/projects/eas-tests-example/builds/5c179d94-ac93-47e7-8ab5-1778726a36ce

### CODE REVIEW

File: app.json [hunk 1]
```diff
@@ -25,10 +25,19 @@
...
     },
     "web": {
       "favicon": "./assets/favicon.png"
+    },
+    "plugins": [
+      "@config-plugins/detox"
+    ],
+    "extra": {
+      "eas": {
+        "projectId": "08f1ceb3-64bf-4718-8f62-c2d372974092"
```
Comment (douglowder) [inline app.json]: I'd suggest removing the `projectId`, and instructing users to clone this repo and run `eas build:configure` and `eas update:configure` to set up the example for their own Expo space.

File: e2e/tests/utils/openAppForDebugBuild.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+const appConfig = require('../../../app.json');
+
+module.exports.openAppForDebugBuild = async function openAppForDebugBuild() {
+  const [platform, target] = process.env.DETOX_CONFIGURATION.split('.');
```
Comment (douglowder) [inline e2e/tests/utils/openAppForDebugBuild.js:4]: What is the advantage of this over `detox.getConfigurationName()`, `detox.getPlatform()`, etc.?

**Review by douglowder — APPROVED** [review]

**Review by brentvatne — APPROVED** [review]

File: e2e/tests/utils/openAppForDebugBuild.js (hunk 2 above)
Comment (dsokal) [reply e2e/tests/utils/openAppForDebugBuild.js:4]: The disadvantage of those is that I didn't know they existed. AFAIR you also didn't used them in your code 😅

Comment (dsokal) [reply e2e/tests/utils/openAppForDebugBuild.js:4]: Those don't exist...

Screenshot 2022-10-11 at 09 20 11
Screenshot 2022-10-11 at 09 19 12

I'm even more confused now.

**Pull request merged**